moon/cfc_77.15.580/sv_net.moon

```diff
@@ -0,0 +1,91 @@
+import Receivers, ReadHeader from net
+import NetworkIDToString from util
+import lower from string
+
+rawget = rawget
+pcall = pcall
+
+import flaggedMessages,
+       netClearTime,
+       netShouldBan,
+       netSpamThreshold,
+       netExtremeSpamThreshold,
+       netSpam,
+       Logger,
+       Webhooker
+       from Section580
+
+sendWebhook = (plySteamID, plyNick, plyIP, identifier, count) ->
+    Webhooker\send "net-spam",
+        steamId: plySteamID
+        name: plyNick
+        ip: plyIP
+        timeframe: netClearTime
+        :identifier
+        :count
+
+net.Incoming = ( len, client ) ->
+    len -= 16
+    strName = NetworkIDToString ReadHeader!
+
+    return unless strName
+
+    plySteamID = "<Unknown Steam ID>"
+    plyNick = "<Unknown Player Name>"
+    plyIP = "<Unknown Player IP>"
+    plyIsValid = IsValid client
+
+    if plyIsValid
+        plySteamID = ply\SteamID!
+        plyNick = ply\Nick!
+        plyIP = ply\IPAddress!
+
+    plyNetSpam = rawget netSpam, plySteamID
+    plyNetSpam or= {}
+    plyNetSpam[strName] or= 0
+    plyNetSpam[strName] += 1
+
+    spamCount = rawget plyNetSpam, strName
+
+    if spamCount > netExtremeSpamThreshold
+        alertMessage = "Player spamming many network messages! #{plyNick} (#{plySteamID}) is spamming: '#{strName}' (Count: #{spamCount} per #{netSpamThreshold} seconds)"
+        Logger\warn alertMessage
+        -- alert staff
+        -- webhooker
+
+        kickReason = "Suspected malicious action"
+
+        if netShouldBan and plyIsValid
+            if ULib
+                ULib.ban client, 1, kickReason
+            else
+                client\Kick kickReason
+
+        return
+
+    if spamCount > netSpamThreshold
+        alertMessage = "Player likely spamming network messages! #{plyNick} (#{plySteamID}) is spamming: '#{strName}' (Count: #{spamCount} per #{netSpamThreshold} seconds)"
+        Logger\warn alertMessage
+        -- alert staff
+        -- webhooker alert
+
+    lowerStr = lower strName
+    func = rawget Receivers, lowerStr
+
+    if not func
+        if not ignoredBadMessages, lowerStr
+            Logger\warn "Nonexistent network message sent by #{plyNick} (#{plySteamID})!: '#{strName}'"
+
+        return
+
+    if rawget flaggedMessages, str
```

Comment:
since network strings get to lowered anyway, in theory someone could skip this check by changing their message case, maybe even automatically, might be better to lower everything and check `lowerStr`?